docs/proposals/service-binding-specification/implementing-specification.md

```diff
@@ -0,0 +1,111 @@
+see-also:
+  - N/A
+replaces:
+  - N/A
+superseded-by:
+  - N/A
+---
+
+# Implementing Service Binding Specification for Kubernetes
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for alpha, beta, GA maturity
+- [ ] User-facing documentation is created in [docs](/docs/)
+
+## Open Questions [optional]
+
+This is where to call out areas of the design that require closure before deciding to implement the
+design. For instance:
+
+> 1. This prevents us from doing cross-namespace binding. Should we do this?
+
+## Summary
+
+This document proposes a mechanism for the Service Binding Operator to implement the
+[Service Binding Specification for Kubernetes](https://github.com/k8s-service-bindings/spec).
+
+## Motivation
+
+The [Service Binding Specification for Kubernetes](https://github.com/k8s-service-bindings/spec)
+aims to create a mechanism that is widely applicable. The benefit of a Kubernetes-wide service binding
+specification is that all of the actors in an ecosystem, including application developers and service
+providers, can work towards a clearly defined abstraction at the edge of their expertise and depend on
+other parties to complete the chain.
+
+On the other hand, the Service Binding Operator has a well-defined mechanism for binding applications
+and services together. This mechanism is different from how the Service Binding Specification for
+Kubernetes lays out binding between entities. Generally, most features defined in the specification
+are a subset of features defined in the Service Binding Operator.
+
+Therefore, allowing users of the Service Binding Operator to choose between the specification and the
+current mechanism is likely to increase the user base of the operator, having a more vibrant community
+of service providers.
+
+### Goals
+
+Goals for this proposal are listed as follows:
+
+- Allow users of the Service Binding Operator to have a clear separation in choosing between the
+specification and the current mechanism for binding applications to services.
+
+### Non-Goals
+
+- Implementation details about how the specification should be implemented.
+
+## Proposal
+
+There are major differences between the specification and the current mechanism the Service Binding
+Operator employes such as how the application is defined in the CRD, how a binding secret is 
+projected in an application container and etc. The [Service Binding feature comparison document](./service-binding-feature-comparison.pdf) lists the differences between the two.
+
+To avoid tangling the implementation between the current codebase and the codebase to support the
+specification, this proposal suggests creating two different channels for the Service Binding
+Operator (e.g. `current` and `spec`) to avoid mixing of implementation details between the two
```

Comment:
Note, OLM is just a mode of packaging and delivering the controller+CRD.Could you help us with an enhancement proposal for the following:  

In an OLM agnostic way,
* How would an existing user of Service Binding migrate their (common) use cases from the existing CRD to the new CRD(s) ?
* How would the project transition its code/structure.

Reply:
thanks for the feedback @sbose78 - those are fair points.  Here are my thoughts:

*  for the purposes of installation, we can probably mirror the OLM channel concept into a non-OLM `install.yaml` (Deployment/Role/RoleBinding/ServiceAccount/CRD) inside somewhere like `deploy/spec-impl`.  That way we're covered in both OLM and non-OLM scenarios.
* the biggest challenge is code co-existence and migration. @navidsh and @arturdzm outlined a nice [comparison](https://github.com/navidsh/service-binding-operator/blob/spec-proposal/docs/proposals/service-binding-specification/feature-comparison.pdf) of features between SBO's `ServiceBindingRequest` (SBR) and the spec's `ServiceBinding` (SB).
  * to maintain maximum stability of the current SBO codebase we could start a new controller and start pulling code from SBO as appropriate - for example, to handle a service provider's annotations, etc.  During this time SBO would have a controller for each of SBR and SB.
  * the next step is for the SB controller to automatically create  (1 or more)`ServiceBinding` CR(s) from a `ServiceBindingRequest` CR, so that customers can still use SBR for a while.  
  * after a certain period of time, SBR is deprecated, and eventually removed.

Reply:
@arthurdm @navidsh

I am curious whether we have a comparison between one-to-one and one-to-many approaches and how the one-to-one approach is expected to work according to your architecture, do you have anything at hand?

Reply:
hey @isutton - we don't have a specific document that clearly describes the 1..1 versus 1..N relationshipps, but that's probably something we should doc in the spec, since we spent quite a bit of time discussing this topic there.  In my mind it boils down to two points which I outlined [here](https://github.com/redhat-developer/service-binding-operator/pull/608#discussion_r482276288).